### PR TITLE
Add validetion and global excpetion hendler

### PR DESCRIPTION
**Pull request by teract10s**

### CODE REVIEW

File: src/main/java/com/example/bookstore/dto/BookDto.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+@Data
+public class BookDto {
+    private String title;
+    private String author;
+    private String isbn;
+    private BigDecimal price;
+    private String description;
+    private String coverImage;
+}
```
Comment (Oleksandr-Tymoshenko) [inline src/main/java/com/example/bookstore/dto/BookDto.java]: you can make it a record

File: src/main/java/com/example/bookstore/dto/CreateBookRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import lombok.Data;
+
+@Data
+public class CreateBookRequestDto {
+    @NotNull
+    private String title;
+    @NotNull
+    private String author;
+    @NotNull
+    private String isbn;
```
Comment (Oleksandr-Tymoshenko) [inline src/main/java/com/example/bookstore/dto/CreateBookRequestDto.java:16]: and for addition you can make regex for isbn

**Review by Oleksandr-Tymoshenko — APPROVED** [review]
Nice, left some minor comments

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Looks like you push both tasks to this PR. Let’s fix it ;) Also, you have merged conflict

File: pom.xml [hunk 3]
```diff
@@ -58,6 +58,15 @@
             <artifactId>mapstruct</artifactId>
             <version>${mapstruct.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-core</artifactId>
+            <version>${liquibase.version}</version>
```
Comment (NazarSmith228) [inline pom.xml]: you don't need to specify liquibase version - it comes from `spring-boot-starter-parent`

File: pom.xml [hunk 4]
```diff
@@ -58,6 +58,15 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.hibernate.validator</groupId>
+            <artifactId>hibernate-validator</artifactId>
+        </dependency>
```
Comment (NazarSmith228) [inline pom.xml]: it is better to use `spring-boot-starter-validation`

File: src/main/java/com/example/bookstore/controller/BookController.java [hunk 5]
```diff
@@ -31,8 +35,24 @@ public BookDto getBookById(@PathVariable Long id) {
     }
 
     @PostMapping("/")
-    public BookDto createBook(@RequestBody CreateBookRequestDto bookDto) {
+    public BookDto createBook(@RequestBody @Valid CreateBookRequestDto bookDto) {
         return bookService.save(bookDto);
     }
 
+    @PutMapping("/{id}")
+    public BookDto updateBook(@PathVariable Long id,
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/controller/BookController.java:48]: just FYI

you may also validate such things as @PathVariable, @RequestParams etc. It looks like:
```
@PathVariable @Positive Long id
```

In order to be able to perform such validation you need to annotate your controller class with `@Validated` annotation

again, this is just FYI - apply it or not :)

File: src/main/java/com/example/bookstore/controller/BookController.java [hunk 6]
```diff
@@ -31,8 +35,24 @@ public BookDto getBookById(@PathVariable Long id) {
...
+                              @RequestBody CreateBookRequestDto createBookRequestDto) {
+        return bookService.updateById(createBookRequestDto, id);
+    }
+
+    @DeleteMapping("/{id}")
+    @ResponseStatus(code = HttpStatus.NO_CONTENT)
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/controller/BookController.java]: how about response statuses for other endpoints?

File: src/main/java/com/example/bookstore/dto/CreateBookRequestDto.java [hunk 7]
```diff
@@ -1,11 +1,21 @@
...
 
-public record CreateBookRequestDto(String title,
-                                   String author,
-                                   String isbn,
-                                   BigDecimal price,
-                                   String description,
-                                   String coverImage) {
+@Data
+public class CreateBookRequestDto {
+    @NotNull
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/dto/CreateBookRequestDto.java]: it is better to use `@NotBlank` with strings

File: src/main/java/com/example/bookstore/dto/CreateBookRequestDto.java [hunk 8]
```diff
@@ -1,11 +1,21 @@
...
+    private String title;
+    @NotNull
+    private String author;
+    @NotNull
+    private String isbn;
+    @NotNull
+    @Min(0)
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/dto/CreateBookRequestDto.java]: `@Min(0)` looks weird :)

could you use `@Positive` or `@PositiveOrZero`?

File: src/main/java/com/example/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
+            MethodArgumentNotValidException ex,
+            HttpHeaders headers,
+            HttpStatusCode status,
+            WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/exception/CustomGlobalExceptionHandler.java]: time is pretty useless in response body - it does not bring any value to the client

status is obvious as well - no sense to have it in body

File: src/main/java/com/example/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+import org.springframework.validation.ObjectError;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@ControllerAdvice
+public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/exception/CustomGlobalExceptionHandler.java:26]: do you have any other exceptions except for `MethodArgumentNotValidException`?

does it make sense to handle them as well?

File: src/main/java/com/example/bookstore/mapper/BookMapper.java [hunk 11]
```diff
@@ -1,14 +1,25 @@
...
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
+        implementationPackage = "<PACKAGE_NAME>.impl"
+)
 public interface BookMapper {
     BookDto toDto(Book book);
 
     Book toBook(CreateBookRequestDto requestDto);
+
+    @BeanMapping(nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
+    void upgradeBookByDto(CreateBookRequestDto requestDto, @MappingTarget Book book);
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/mapper/BookMapper.java]: let's simplify it :)
```suggestion
    void updateBook(CreateBookRequestDto requestDto, @MappingTarget Book book);
```

File: src/main/java/com/example/bookstore/repository/BookRepository.java [hunk 12]
```diff
@@ -1,13 +1,8 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
 
-public interface BookRepository {
-    Book save(Book book);
-
-    List<Book> findAll();
-
-    Optional<Book> findById(Long id);
+public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/repository/BookRepository.java:9]: `@Repository` annotation is missing

File: src/main/java/com/example/bookstore/repository/BookSpecificationBuilder.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+        }
+        return spec;
+    }
+
+    private static Specification<Book> getTitleSpecification(String[] titles) {
+        return (root, query, criteriaBuilder) -> root.get("title")
+                .in(Arrays.stream(titles).toArray());
+    }
+
+    private static Specification<Book> getAuthorSpecification(String[] authors) {
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/repository/BookSpecificationBuilder.java]: there is no sense to make these methods `static` :)

File: src/main/resources/liquibase.properties [hunk 14]
```diff
@@ -0,0 +1,5 @@
+url=jdbc:h2:mem:testdb
+username=root
+password=1234
+changeLogFile=db/changelog/db.changelog-master.yaml
+driver=org.h2.Driver
```
Comment (NazarSmith228) [inline src/main/resources/liquibase.properties:5]: in your app config you're using MySQL, however in liquibase config - H2

why?)

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/resources/liquibase.properties (hunk 14 above)
Comment (teract10s) [reply src/main/resources/liquibase.properties:5]: No, but my project can't be builded in github if I leave the settings for mysql. We talked about it with Lilia on QnA and she said to leave it like that for now

**Pull request merged**